src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -130,7 +126,17 @@ public async ValueTask SynchronizeAssetsAsync(HashSet<Checksum> checksums, Cance
 
             using (Logger.LogBlock(FunctionId.AssetService_SynchronizeAssetsAsync, Checksum.GetChecksumsLogInfo, checksums, cancellationToken))
             {
-                var checksumsArray = checksums.ToImmutableArray();
+                using var _1 = ArrayBuilder<Checksum>.GetInstance(checksums.Count, out var missingChecksums);
+                foreach (var checksum in checksums)
+                {
+                    if (checksum == Checksum.Null)
+                        continue;
+
+                    if (!_assetCache.TryGetAsset<object>(checksum, out _))
```

Comment:
moved this check from callsites *into* the asset provider.  it now ensures that it doesn't request anything if it already has it (Regardless of what the caller passed in).